### PR TITLE
Feature Added: es6 maps support

### PR DESCRIPTION
**Pull request by keenondrums**

### CODE REVIEW

**Comment by keenondrums** [comment]
It covers only ES6 Maps. I'm going to work on ES6 Sets, but Sets will have certain design challenges, so I would really really like to leave them for a separate PR.

File: __tests__/patch.js [hunk 1]
```diff
@@ -118,6 +150,30 @@ describe("delete 1", () => {
     )
 })
 
+describe("delete 2", () => {
+    runPatchTest(
+        new Map([["x", 1]]),
+        d => {
+            d.delete("x")
+        },
+        [{op: "remove", path: ["x"]}],
```
Comment (aleclarson) [inline __tests__/patch.js]: Are `Map` entries assumed to be JSON-compatible?

File: src/common.js [hunk 2]
```diff
@@ -30,16 +31,36 @@ export function original(value) {
...
+        for (let key in override) {
+            if (has(override, key)) {
+                target[key] = override[key]
+            }
+        }
+    })
+    return target
+}
+export function assign(target) {
+    const overrides = Array.prototype.slice.call(arguments, 1)
```
Comment (aleclarson) [inline src/common.js]: You can do `assign(target, ...overrides)` here

File: __tests__/patch.js (hunk 1 above)
Comment (keenondrums) [reply __tests__/patch.js]: I guess one could serialize a Map to an object if needed. So not, not out-of-the-box. but yes with certain additional work.

File: src/common.js [hunk 3]
```diff
@@ -30,16 +31,36 @@ export function original(value) {
...
+    if (isMap(target)) {
+        return assignMap(target, overrides)
     }
+    if (Object.assign) {
+        return Object.assign.apply(this, Array.prototype.slice.call(arguments))
```
Comment (aleclarson) [inline src/common.js]: This should be `Object.assign(target, ...overrides)`

File: src/common.js [hunk 4]
```diff
@@ -80,11 +102,11 @@ export function shallowCopy(base, invokeGetters = false) {
 }
 
 export function each(value, cb) {
-    if (Array.isArray(value)) {
-        for (let i = 0; i < value.length; i++) cb(i, value[i], value)
-    } else {
-        ownKeys(value).forEach(key => cb(key, value[key], value))
+    if (Array.isArray(value) || isMap(value)) {
+        value.forEach((valueI, index) => cb(index, valueI, value))
```
Comment (aleclarson) [inline src/common.js]: nit: Maybe `entry` instead of `valueI`?

File: src/common.js (hunk 2 above)
Comment (keenondrums) [reply src/common.js]: Wouldn't it fail for ES5 environments?

File: src/patches.js [hunk 5]
```diff
@@ -1,9 +1,28 @@
-import {each} from "./common"
+import {each, isMap} from "./common"
 
 export function generatePatches(state, basePath, patches, inversePatches) {
-    Array.isArray(state.base)
-        ? generateArrayPatches(state, basePath, patches, inversePatches)
-        : generateObjectPatches(state, basePath, patches, inversePatches)
+    let generatePatchesFn = generatePatchesFromAssigned(
```
Comment (aleclarson) [inline src/patches.js]: The flow should be:
```js
generatePatchesFn =
  isArray ? generateArrayPatches
  : isMap ? ...
  // isObject
  : ...
```
And please hoist the functions passed to `generatePatchesFromAssigned` so they can be reused. 👍

File: src/patches.js [hunk 6]
```diff
@@ -1,9 +1,28 @@
...
+            return key in obj
+        }
+    )
+    if (Array.isArray(state.base)) {
+        generatePatchesFn = generateArrayPatches
+    }
+    if (isMap(state.base)) {
+        generatePatchesFn = generatePatchesFromAssigned(
+            function(map, key) {
+                return map.get(key)
```
Comment (aleclarson) [inline src/patches.js]: These functions would look cleaner with arrow functions

File: src/patches.js [hunk 7]
```diff
@@ -103,29 +128,55 @@ export function applyPatches(draft, patches) {
...
+                    return
+                }
+                base[key] = value
+            }
+            function add(key, value) {
+                if (isMap(base.base)) {
+                    base.set(key, value)
+                    return
+                }
+                if (Array.isArray(base)) {
```
Comment (aleclarson) [inline src/patches.js]: nit: Probably want `isArray` check before `isMap` since it's more likely to be true

File: src/patches.js [hunk 8]
```diff
@@ -103,29 +128,55 @@ export function applyPatches(draft, patches) {
...
+                    return
+                }
+                base[key] = value
+            }
+            function remove(key) {
+                if (isMap(base.base)) {
+                    base.delete(key)
+                    return
+                }
+                if (Array.isArray(base)) {
```
Comment (aleclarson) [inline src/patches.js]: Same here

File: src/proxy.js [hunk 9]
```diff
@@ -39,11 +40,17 @@ export function createProxy(base, parent) {
...
-          // although state itself is an object
-          Proxy.revocable([state], arrayTraps)
-        : Proxy.revocable(state, objectTraps)
+    let proxyTarget = state
+    let traps = objectTraps
+    if (Array.isArray(base)) {
+        proxyTarget = [state]
+        traps = arrayTraps
+    }
+    if (isMap(base)) {
```
Comment (aleclarson) [inline src/proxy.js]: Should be `else if`

File: src/common.js (hunk 2 above)
Comment (aleclarson) [reply src/common.js]: Immer is compiled to ES5 by [bili](https://github.com/egoist/bili), so no. ;)

**Comment by keenondrums** [comment]
@aleclarson addressed your comments. Thanks for the prompt review!
Currently, the build is failing because of the generator function. Working on it.

**Comment by keenondrums** [comment]
@aleclarson I made Buble pass generators through, so you could see all the tests passing. Though it's still not a solution. Makes me think if nobody could think of a better way to wrap iterators for Maps, `immer` would have to go back to doing separate build for ES5 envs. Or not to support Maps.
What do you think?

**Comment by aleclarson** [comment]
I'm working on changes that will include a fix to the generator issue. Then we can publish this!

**Comment by keenondrums** [comment]
@aleclarson are thinking of implementing iterator protocol yourself?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols

**Comment by aleclarson** [comment]
Here's my work so far: https://github.com/immerjs/immer/commits/v3.1

There is a strange recursion bug that I'm trying to track down. Could you take a look?

**Comment by keenondrums** [comment]
Will be pleased to do it tomorrow. It's 1AM where I am :)

**Comment by keenondrums** [comment]
Closing it in favor of https://github.com/immerjs/immer/pull/351

**Pull request closed without merging**